File: tests/login_test.rs
use dcss_api::{Error, Webtile};
use serde_json::{from_str, Value};

#[test]
fn successful_credential_login() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let _ = webtile
        .login_with_credentials("Username", "Password")
        .expect("Login failed.");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn multiple_login_same_user() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let _ = webtile
        .login_with_credentials("Username", "Password")
        .expect("Login failed.");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let _ = webtile
        .login_with_credentials("Username", "Password")
        .expect("Login failed.");

    let mut found = 0;
    while let Some(message) = webtile.get_message() {
        let message_obj = message.as_object().unwrap();
        if message_obj["msg"] == "login_success" {
            found = 1;
            let json: Value = from_str(
                "{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}",
            )
            .unwrap();
            assert_eq!(json, message);
        }
    }
    if found == 0 {
        panic!("Failed to find a login success.")
    }

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn multiple_login_diff_user() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let _ = webtile
        .login_with_credentials("Username", "Password")
        .expect("Login failed.");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let _ = webtile
        .login_with_credentials("Username2", "Password")
        .expect("Login failed.");

    let mut found = 0;
    while let Some(message) = webtile.get_message() {
        let message_obj = message.as_object().unwrap();
        if message_obj["msg"] == "login_success" {
            found = 1;
            let json: Value = from_str(
                "{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username2\"}",
            )
            .unwrap();
            assert_eq!(json, message);
        }
    }
    if found == 0 {
        panic!("Failed to find a login success.")
    }

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn failed_credential_login() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let result = webtile.login_with_credentials("XXXXXXX", "XXXXXXX");

    assert!(matches!(result, Err(Error::LoginFailed)));

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn failed_credential_login_and_retry() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let result = webtile.login_with_credentials("XXXXXXX", "XXXXXXX");
    assert!(matches!(result, Err(Error::LoginFailed)));

    while webtile.get_message().is_some() {}

    let _ = webtile
        .login_with_credentials("Username", "Password")
        .unwrap();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn get_cookie_and_login() {
    // Connect to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Log in (to a user called "Username", with a password "Password")
    let _ = webtile
        .login_with_credentials("Username", "Password")
        .expect("Failed to login");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    // Get cookie from the game
    let cookie = webtile.request_cookie().unwrap();

    assert_eq!("Username%", &cookie[0..9]);

    // Disconnect from DCSS Webtile
    webtile.disconnect().expect("Failed to disconnect.");

    // Connect (again) to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Login with cookie
    let _ = webtile
        .login_with_cookie(cookie.as_str())
        .expect("Failed to login");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn failed_cookie_login() {
    // Connect to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Login with cookie
    let result = webtile.login_with_cookie("Username%123456789123456789123456789");
    assert!(matches!(result, Err(Error::LoginFailed)));

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn using_old_cookie_login() {
    // Connect to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Log in (to a user called "Username", with a password "Password")
    let _ = webtile
        .login_with_credentials("Username", "Password")
        .expect("Failed to login");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    // Get cookie from the game
    let first_cookie = webtile.request_cookie().unwrap();

    assert_eq!("Username%", &first_cookie[0..9]);

    // Disconnect from DCSS Webtile
    webtile.disconnect().expect("Failed to disconnect.");

    // Connect to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Login with cookie
    let _ = webtile
        .login_with_cookie(first_cookie.as_str())
        .expect("Failed to login.");

    let _ = webtile.get_message();

    let json: Value =
        from_str("{\"admin\": false, \"msg\": \"login_success\", \"username\": \"Username\"}")
            .unwrap();
    assert_eq!(Some(json), webtile.get_message());

    // Get cookie from the game
    let second_cookie = webtile.request_cookie().unwrap();

    assert_eq!("Username%", &second_cookie[0..9]);

    // Disconnect from DCSS Webtile
    webtile.disconnect().expect("Failed to disconnect.");

    // Connect to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Login with cookie
    let result = webtile.login_with_cookie(first_cookie.as_str());

    assert!(matches!(result, Err(Error::LoginFailed)));

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn credential_login_gameid() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let gameid = webtile
        .login_with_credentials("Username", "Password")
        .expect("Login failed.");

    println!("{:?}", gameid);

    let test_gameid = vec![
        "dcss-web-trunk".to_owned(),
        "seeded-web-trunk".to_owned(),
        "tut-web-trunk".to_owned(),
        "sprint-web-trunk".to_owned(),
    ];
    assert_eq!(gameid, test_gameid);

    webtile.disconnect().expect("Failed to disconnect");
}

#[test]
fn cookie_login_gameid() {
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect.");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    let gameid = webtile
        .login_with_credentials("Username", "Password")
        .expect("Login failed.");

    let test_gameid = vec![
        "dcss-web-trunk".to_owned(),
        "seeded-web-trunk".to_owned(),
        "tut-web-trunk".to_owned(),
        "sprint-web-trunk".to_owned(),
    ];
    assert_eq!(gameid, test_gameid);

    // Get cookie from the game
    let cookie = webtile.request_cookie().unwrap();

    assert_eq!("Username%", &cookie[0..9]);

    // Disconnect from DCSS Webtile
    webtile.disconnect().expect("Failed to disconnect.");

    // Connect (again) to DCSS Webtile
    let mut webtile =
        Webtile::connect("ws://localhost:8080/socket", 0, "0.29").expect("Failed to connect");

    // Empty message queue;
    while webtile.get_message().is_some() {}

    // Login with cookie
    let gameid = webtile
        .login_with_cookie(cookie.as_str())
        .expect("Failed to login");

    let test_gameid = vec![
        "dcss-web-trunk".to_owned(),
        "seeded-web-trunk".to_owned(),
        "tut-web-trunk".to_owned(),
        "sprint-web-trunk".to_owned(),
    ];
    assert_eq!(gameid, test_gameid);

    webtile.disconnect().expect("Failed to disconnect");
}

File: python/tests/login_test.py
import pytest
import dcss_api
from dcss_api import APIErr
import json


def test_successful_credential_login():
    # Connect to DCSS Webtile
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    webtile.disconnect()


def test_multiple_login_same_user():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    found = False
    while (message := webtile.get_message()) != None:
        json_message = json.loads(message)
        if json_message["msg"] == "login_success":
            base_message = '{"admin":false,"msg":"login_success","username":"Username"}'
            if base_message == message:
                found = True

    assert found

    webtile.disconnect()


def test_multiple_login_diff_user():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username2", "Password")

    found = False
    while (message := webtile.get_message()) != None:
        json_message = json.loads(message)
        if json_message["msg"] == "login_success":
            base_message = (
                '{"admin":false,"msg":"login_success","username":"Username2"}'
            )
            if base_message == message:
                found = True

    assert found

    webtile.disconnect()


def test_failed_credential_login():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    try:
        webtile.login_with_credentials("XXXXXXX", "XXXXXXX")
        assert False
    except APIErr as e:
        if "Failed to login" in e.args[0]:
            assert True
        else:
            assert False

    webtile.disconnect()


def test_failed_credential_login_and_retry():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    try:
        webtile.login_with_credentials("XXXXXXX", "XXXXXXX")
        assert False
    except APIErr as e:
        if "Failed to login" in e.args[0]:
            assert True
        else:
            assert False

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    webtile.disconnect()


def test_get_cookie_and_login():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    cookie = webtile.request_cookie()

    assert "Username%" == cookie[0:9]

    webtile.disconnect()

    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_cookie(cookie)

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    webtile.disconnect()


def test_failed_cookie_login():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    try:
        webtile.login_with_cookie("Username%123456789123456789123456789")
        assert False
    except APIErr as e:
        if "Failed to login" in e.args[0]:
            assert True
        else:
            assert False


def test_using_old_cookie_login():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_credentials("Username", "Password")

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    first_cookie = webtile.request_cookie()

    assert "Username%" == first_cookie[0:9]

    webtile.disconnect()

    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    webtile.login_with_cookie(first_cookie)

    webtile.get_message()

    assert (
        '{"admin":false,"msg":"login_success","username":"Username"}'
        == webtile.get_message()
    )

    second_cookie = webtile.request_cookie()

    assert "Username%" == second_cookie[0:9]

    webtile.disconnect()

    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    try:
        webtile.login_with_cookie(first_cookie)
        assert False
    except APIErr as e:
        if "Failed to login" in e.args[0]:
            assert True
        else:
            assert False

    webtile.disconnect()


def test_credential_login_gameid():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    gameid = webtile.login_with_credentials("Username", "Password")

    assert gameid == [
        "dcss-0.29",
        "dcss-0.30",
        "dcss-0.31",
        "dcss-0.32",
    ]

    webtile.disconnect()


def test_cookie_login_gameid():
    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    gameid = webtile.login_with_credentials("Username", "Password")

    assert gameid == [
        "dcss-0.29",
        "dcss-0.30",
        "dcss-0.31",
        "dcss-0.32",
    ]

    cookie = webtile.request_cookie()

    assert "Username%" == cookie[0:9]

    webtile.disconnect()

    webtile = dcss_api.WebtilePy("ws://localhost:8080/socket", 0, "0.32")

    while (message := webtile.get_message()) != None:
        pass

    gameid = webtile.login_with_cookie(cookie)

    assert gameid == [
        "dcss-0.29",
        "dcss-0.30",
        "dcss-0.31",
        "dcss-0.32",
    ]

    webtile.disconnect()
